### ISSUE TITLE
High DPI: Position of checkboxes shown in VT is not optimal

### ISSUE THREAD

**Issue by Ajoschi** [JAM-Software/Virtual-TreeView#585]
Use a VirtualStringTree with check box support (TVTMiscOption.toCheckSupport included in TreeOptions.MiscOptions) and run your application on a machine with a resolution or 150/200%
-> Checkboxes are not shown centered (like e.g. the images), but on the upper left corner.

Since the VirtualStringTree behaves very well on all other aspects of High DPI awareness, it would be very nice if this one could also be resolved =)

Thanks in advance,
Ajoschi

**Comment by joachimmarder** [comment]
Häng doch bitte noch den Screenshot an, denn die Issues dienen als Doku für
die Commits.

Danke & LG ...

Joey


2015-10-07 15:21 GMT+02:00 Ajoschi <notifications@github.com>:

**Comment by joachimmarder** [comment]
Please attach a screenshot.

**Comment by Ajoschi** [comment]
![virtualstringtree_checkbox_highdpi](https://cloud.githubusercontent.com/assets/15014911/10359369/2efc4322-6d98-11e5-962e-3865b72c5db9.png)

**Comment by studylamp** [comment]
Please note that this is not just a placement issue.  The size of the checkboxes do not scale either.  In fact, on a 200% scaled screen (the default on the Surface Book, for example), the checkboxes are very difficult to read and use because they are so small, regardless of placement.

**Comment by joachimmarder** [comment]
You are right, I noticed that too. This is why this cannot be fixed quickly and it may take some more days.

**Comment by joachimmarder** [comment]
With the recent changes, it looks better now. However, at 200% Windows scaling, a VCL checkbox has 26x26 pixels, and a Virtual TreeView checkbox has 31x31 pixels and therefore looks pixelated.

**Comment by IgitBuh** [comment]
I've updated from an older VTV version to the latest 6.2.3 mainly for fixing DPI scaling issues. I'd like to bring to your attention that the checkboxes still aren't resized correctly in my projects. I'm using ckSystemDefault and ckDarkTick, but none of them is scaled better than the old version I've used before.

Do I need to make further adjustments to make your fix work? I assume it should work without any additional changes. I'd include a screenshot, but it looks exactly the same as in the screenshot posted by Ajoschi.

Nobody here confirmed that the fix worked for them. Is this fix really a _closed_ bug?

**Edit:** I've now tried out all the other CheckImageKind options and noticed that with ckSystemFlat it does get sized correctly. However, it's a totally different UI element than all the other checkboxes that are used in the application so I'd rather not use it. So what about correct DPI scaling for ckSystemDefault (and the others like ckDarkTick)?

**Comment by joachimmarder** [comment]
> Is this fix really a closed bug?

I was able to replicate the original issue, it applied to the system checkbox images from the active Windows theme, which are used if the toThemeAware flag is set. This issue is fixed.

> However, it's a totally different UI element than all the other checkboxes

This applies to all checkbox images that do not come from the active style (which can be the Windows style or a VCL style). `ckLightCheck .. ckXP` are loaded from 16x16 pixel resource bitmaps, so they are static (not theme aware) and won't scale well. `ckSystemFlat` and `ckSystemDefault` are generated using the API call `DrawFrameControl()`, I think these are the checkbox images of the Windows classic style (the Windows 2000 style). So I guess on Windows XP with classic style activated these checkbox images would fit the others you see  ;-) But at least they scale quite well because they are not static but generated at runtime.

I personally only use Virtual TreeView controls with the toThemeAware option. With having Windows styles and VCL styles I don't see much need for checkbox images in custom styles and considered removing the `CheckImageKind` property. Maybe I will open a new issue for that idea to allow discussion.

**Comment by IgitBuh** [comment]
I'm sorry, I don't get it. toThemeAware is enabled by default and so is it in my projects. With it being enabled the checkboxes are not scaled. After reading your comment I've disabled toThemeAware just for a quick test. I noticed that the checkboxes now look Win2000ish but now they're scaled properly (sharp)! So what's going on here? According to your comment it should be the other way around. I'm testing this on Windows 10 with 175% DPI scaling. **Edit:** Same result with Windows 7 and 150%.

As a side note: toUseExplorerTheme sounds like a duplicate of toThemeAware. I understand the difference in practice but the naming is confusing for new users.

Side note 2: Removing the property CheckImageKind isn't a good idea in my optionion as it would break many project without need. There are good reasons especially for chCustom.

**Comment by joachimmarder** [comment]
Maybe posting a screenshot would be good. And please post which RAD Studio version you are using. My comments apply to XE8.

> There are good reasons especially for chCustom.

Support for a custom image list should stay of course, but for that you only need the property `CustomCheckImages`.

> Removing the property CheckImageKind isn't a good idea in my optionion as it would break many project without need.

In case they were not using toThemeAware or custom checkbox images anyway, they would fallback to the default checkbox images of the active style. Doesn't sound that bad to me. But we should discuss this in the new issue #622 .

**Comment by IgitBuh** [comment]
It looks exactly the same like in the screenshot Ajoschi posted. If it really helps I can prepare some additional screenshots, of course.
I'm using Delphi XE (sglienke fork). Is this really relevant in this case?

**Comment by IgitBuh** [comment]
![vtv-623-dpi-checkboxes-2](https://cloud.githubusercontent.com/assets/703571/15000356/8d93e2aa-118f-11e6-8977-f388a45847d6.png)

**Comment by joachimmarder** [comment]
I'm sorry, but I don't use or support Delphi XE and so cannot tell you if it is relevant.
I addressed this issue using the Advanced sample project, module "General abilities Demo", Delphi XE8 and Windows 10 64Bit. I saw what Ajoschi reported and it looked much better after the fix.

**Comment by Ajoschi** [comment]
Hi IgitBuh.

I'm using Delphi XE7. toThemeAware is activated and CheckImageKind is ckSystemDefault.
All looks fine in my application.

**Comment by IgitBuh** [comment]
I've just finished reading through all of the XE related topics here. I don't want to start a new discussion about this, but I'd like to mention my sincere sorrow about the fact that VTV is willingly leaving a huge user base behind by burning bridges to older versions. After all: XE3 is merely 3,5 years old and upgrades are too expensive for most developers who don't really need the new features.

I don't know how this issue might be related to the Delphi version, but I'll try to investigate. Unfortunately this topic isn't exactly my area of expertise.

**Comment by joachimmarder** [comment]
> Unfortunately this topic isn't exactly my area of expertise.

The same applies for me regarding the high dpi capabilities of Delphi XE.

I know that this is in general working OK in XE5 - XE8, and was further improved in 10 and 10.1. I just upgraded to 10.1 and see that the changes in VCL may require some additional work and may $ifdefs in Virtual TreeView for proper scaling.

**Comment by IgitBuh** [comment]
I compared the XE-forked version to the official version.
Seems like the checkbox calculation has been altered there.
Original:

```
StyleServices.GetElementSize(Canvas.Handle, Details, TElementSize.esActual, lSize);
R := Rect(XPos, YPos, XPos + lSize.cx, YPos + lSize.cy);
```

sglienke fork:

```
{$ifdef SUPPORTS_VCL_STYLES}
      if not StyleServices.GetElementSize(Canvas.Handle, Details, TElementSize.esActual, lSize) then begin
        // radio buttons fail in RAD Studio 10 Seattle and lower, fallback to checkbox images. Siee issue #615
        if not StyleServices.GetElementSize(Canvas.Handle, StyleServices.GetElementDetails(tbCheckBoxUncheckedNormal), TElementSize.esActual, lSize) then
          lSize := TSize.Create(GetSystemMetrics(SM_CXMENUCHECK), GetSystemMetrics(SM_CYMENUCHECK));
      end;//if
      R := Rect(XPos, YPos, XPos + lSize.cx, YPos + lSize.cy);
{$else}
      R := Rect(XPos - 1, YPos + 1, XPos + 16, YPos + 16);
{$endif}
```

XE applies to the else-case, so the checkbox size is hardcoded there.
I was under the impression that `GetSystemMetrics(SM_CXMENUCHECK)` will easily get the checkbox size, but apparently that's not the case. At least not for themed checkboxes.

I went through all common DPI scaling factors and measured the actual values. If anybody's interested, here they are:

```
100%: 15px (SM_CXMENUCHECK: 15)
125%: 16px (SM_CXMENUCHECK: 19)
150%: 20px (SM_CXMENUCHECK: 25)
175%: 20px (SM_CXMENUCHECK: 29)
200%: 26px (SM_CXMENUCHECK: 31)
250%: 32px (SM_CXMENUCHECK: 41)
300%: 40px (SM_CXMENUCHECK: 47)
```

I don't see any correlation between the actual size and the scaling factor. So to bring this to an end, I'll include a case-statement with those measured values hardcoded into VTV. While hardcoded values are never a good idea, I don't see another quick solution here (other than invisibly painting the checkbox into a bitmap and measuring it - which I don't want to do for performance reasons) and it's definitely better than the current solution.

Is there a way to bring this to the attention of sglienke who might want to include my final solution into the his fork? I'm sorry for my novice question, but I'm not familiar with github nor offers his fork an issue tracker.

**Comment by IgitBuh** [comment]
Whoever's interested, here's my solution for Delphi XE:

```
//      R := Rect(XPos - 1, YPos + 1, XPos + GetSystemMetrics(SM_CYMENUCHECK), YPos + GetSystemMetrics(SM_CYMENUCHECK));
      // Workaround, because Delphi XE doesn't provide StyleServices.GetElementSize
      // Alternatively use JVCL/JvThemes/StyleServices.GetElementSize
      case Screen.PixelsPerInch of
        96:  checkboxSize := 15; //100% (15)
        120: checkboxSize := 16; //125% (19)
        144: checkboxSize := 20; //150% (25)
        168: checkboxSize := 20; //175% (29)
        192: checkboxSize := 26; //200% (31)
        240: checkboxSize := 32; //250% (41)
        288: checkboxSize := 40; //300% (47)
        else checkboxSize := GetSystemMetrics(SM_CXMENUCHECK);
      end;
      marginX := -1;
      marginY := (GetSystemMetrics(SM_CXMENUCHECK) - checkboxSize) div 2; //MulDiv(1, Screen.PixelsPerInch, 96);
      R := Rect(XPos + marginX, YPos + marginY, XPos + checkboxSize + marginX, YPos + checkboxSize + marginY);
```

I'm not fully satisfied with marginY, because actually I'd need the node height but I don't seem to be able to retrieve it at that point. The workaround is fine for a normal hode height, though. Actually this also concerns the official VTV version, because there's no calculation of the Y position included in the official code.

Other than that this solves the problem for me. I'd appreciate it if sglienke could include it into his fork.

**Comment by joachimmarder** [comment]
> `{$ifdef SUPPORTS_VCL_STYLES}`
> [...]
> {$else}
> [...]
> XE applies to the else-case, so the checkbox size is hardcoded there.
> I went through all common DPI scaling factors and measured the actual values. 
> [...]
> I don't see any correlation between the actual size and the scaling factor.

This is exactly the crap why I don't want to support XE any more. Maintaining an XE-only fork would be much easier for someone who is using this daily. But doing XE - 10.1 in a single source is annoying and extremely time consuming for a UI component.

> nor offers his fork an issue tracker.

In another post Stefan made clear that he won't open an issue tracker. You could of course try a pull request.

**Comment by joachimmarder** [comment]
> Actually this also concerns the official VTV version, because there's no calculation of the Y position included in the official code.

I just opened issue #624 for this. Does the screenshot show what you mean?

**Comment by IgitBuh** [comment]
I understand your reasons and while I was writing my previous two comments I already felt you'd point that out. All I'm saying is that by doing so you're leaving a big user base behind. It's like when some drivers aren't compatible with a new Windows version. You might say it's for a good reason, but in reality people are still using XP mainly because they don't want to or even can't throw away their already existing hardware like printers, scanners, industry related devices etc.

**Comment by IgitBuh** [comment]
> I just opened issue #624 for this. Does the screenshot show what you mean?

Probably yes. In short: The checkbox drawing method currently isn't aware of the node height. So the checkbox (and probably radio buttons like in your screenshot) only look correctly vertically aligned if the node height equals the checkbox height.

**Comment by joachimmarder** [comment]
Another just fixed high dpi issue with 10.1 Berlin: issue #625
